### PR TITLE
Skip and warn of duplicated SQL Server foreign keys

### LINKED ISSUE #25197: Unblocking duplicate foreign keys when reverse engineering an existing database (SQL Server)
I frequently see this error when users of SQL Server databases reverse engineer an existing database.

One possible solution could be to let the SQL Server scaffolding factory ignore the "second" foreign key. (and maybe just log a warning)

Thoughts?

```
System.InvalidOperationException: The foreign key {'Activity'} cannot be added to the entity type 'MyChild' because a foreign key on the same properties already exists on entity type 'MyChild' and also targets the key {'Oid'} on 'MyParent'.
   at void Microsoft.EntityFrameworkCore.Metadata.Internal.ForeignKey.Validate(IReadOnlyList<Property> properties, Key principalKey, EntityType declaringEntityType, EntityType principalEntityType)
   at new Microsoft.EntityFrameworkCore.Metadata.Internal.ForeignKey(IReadOnlyList<Property> dependentProperties, Key principalKey, EntityType dependentEntityType, EntityType principalEntityType, ConfigurationSource configurationSource)
   at ForeignKey Microsoft.EntityFrameworkCore.Metadata.Internal.EntityType.AddForeignKey(IReadOnlyList<Property> properties, Key principalKey, EntityType principalEntityType, ConfigurationSource? componentConfigurationSource, ConfigurationSource configurationSource)
   at IMutableForeignKey Microsoft.EntityFrameworkCore.Metadata.Internal.EntityType.Microsoft.EntityFrameworkCore.Metadata.IMutableEntityType.AddForeignKey(IReadOnlyList<IMutableProperty> properties, IMutableKey principalKey, IMutableEntityType principalEntityType)
   at IMutableForeignKey Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitForeignKey(ModelBuilder modelBuilder, DatabaseForeignKey foreignKey)
   at ModelBuilder Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitForeignKeys(ModelBuilder modelBuilder, IList<DatabaseForeignKey> foreignKeys)
   at ModelBuilder Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitDatabaseModel(ModelBuilder modelBuilder, DatabaseModel databaseModel)
   at IModel Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.Create(DatabaseModel databaseModel, ModelReverseEngineerOptions options)

```

EF Core version: 5, 6
Database provider: often Microsoft.EntityFrameworkCore.SqlServer

### PR DESCRIPTION
**Pull request by ErikEJ**
fixes #25197

### CODE REVIEW

**Comment by ErikEJ** [comment]
Wauw, the logging is complex and so full of magic stuff.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 1]
```diff
@@ -1214,6 +1214,16 @@ private void GetForeignKeys(DbConnection connection, IReadOnlyList<DatabaseTable
                         }
                         else
                         {
+                            var duplicated = table.ForeignKeys.FirstOrDefault(k => k.Columns.SequenceEqual(foreignKey.Columns));
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs:1217]: The FK should also have same principal table for duplicate. Multiple overlapping FK targeting different entity types should work.

Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs:1217]: Wauw, never thought that was possible with SQL Server!

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 2]
```diff
@@ -1214,6 +1214,18 @@ private void GetForeignKeys(DbConnection connection, IReadOnlyList<DatabaseTable
                         }
                         else
                         {
+                            var duplicated = table.ForeignKeys
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs:1217]: match schema also? 🤔

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 1 above)
Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Fixed

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 2 above)
Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs:1217]: Principal table contains both name and schema - not sure what you mean??

Comment (smitpatel) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs:1217]: Didn't realize PrincipalTable was class. Thought it was string.

File: test/EFCore.SqlServer.FunctionalTests/Scaffolding/SqlServerDatabaseModelFactoryTest.cs [hunk 3]
```diff
@@ -2359,6 +2359,45 @@ public void Skip_reflexive_foreign_key()
...
+CREATE TABLE OtherPrincipalTable (
+    Id int PRIMARY KEY,
+);
+
+CREATE TABLE DependentTable (
+    Id int PRIMARY KEY,
+    ForeignKeyId int,
+    CONSTRAINT MYFK1 FOREIGN KEY (ForeignKeyId) REFERENCES PrincipalTable(Id),
+    CONSTRAINT MYFK2 FOREIGN KEY (ForeignKeyId) REFERENCES PrincipalTable(Id),
+	CONSTRAINT MYFK3 FOREIGN KEY (ForeignKeyId) REFERENCES OtherPrincipalTable(Id),
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Scaffolding/SqlServerDatabaseModelFactoryTest.cs:2379]: Nit: Just align this line.

**Review by smitpatel — APPROVED** [review]

**Comment by ErikEJ** [comment]
@bricelam any chance to get this into 6.0??

**Comment by smitpatel** [comment]
@ErikEJ - This will get into 6.0. I will merge this by Wednesday (or earlier) so it is in next preview version.

**Comment by smitpatel** [comment]
Thanks @ErikEJ

**Pull request merged**